### PR TITLE
missing-transactional lint; extrinsics-must-be-tagged lint; lintcheck integration test

### LINKED ISSUE #10: Add the missing-transactional lint for detecting extrinsics without the #[transactional] marker
Extrinsics can be optionally labeled with `#[transactional]` in Substrate, but we believe that should be the default. We want a lint that ensures all extrinsics are marked as `#[transactional]`.

### LINKED ISSUE #13: Add the extrinsics-must-be-tagged lint and subcommand to detect extrinsic reorderings
Currently, extrinsics should never be reordered, unless we add an explicit tag to ensure that the serialization remains the same. We need a lint that forces this tag to be added.

Untagged extrinsic: https://github.com/ComposableFi/composable/blob/4e1ca505a59d914cc37d3571f0aefa1eb7cf2218/code/parachain/frame/assets/src/lib.rs#L178

### PR DESCRIPTION
**Pull request by TheCodingWombat**
- [x] Create the missing-transactional lint that flags extrinsics that do not use the `with_transaction` function.
- [x] Create the extrinsics-must-be-tagged lint that flags extrinsics that are not tagged with the `pallet::call_index` macro.
- [x] Auto-fix for lint missing-transactional and extrinsics-must-be-tagged
- [x] Fix broken auto-fix for no-panics lint
- [x] Update readme to list current and upcoming lints.
- [x] Lintcheck (forked from rust-clippy) to run substrace on a defined set of crates and compare lint output to expected warnings/errors.

Integration testing framework (for these and future lints) will be in a different PR. For now test by running on substrate-node-template/pallets/template.

Closes #10 
Closes #13

### CODE REVIEW

File: .github/workflows/build_test_clippy.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    - uses: dtolnay/rust-toolchain@master
+      with:
+        toolchain: nightly-2022-11-21
+        components: clippy
+    - name: Build
+      run: cargo build --verbose
+    - name: Run tests
+      run: cargo test --verbose
+    - name: Run clippy
+      run: cargo clippy --verbose
```
Comment (KaiserKarel) [inline .github/workflows/build_test_clippy.yml]: Has this changed vs what's in main?

File: Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,8 @@ semver = "1.0"
 rustc_tools_util = "0.2.0"
 tempfile = { version = "3.2", optional = true }
 termize = "0.1"
+syn = { version = "1.0", features = ["full"] } # are these necessary?
+quote = "1.0" # TODO: Are these necessary?
```
Comment (KaiserKarel) [inline Cargo.toml]: Remove and see if compiles?

File: substrace_lints/src/substrace_lints/extrinsics_must_be_tagged.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                "Extrinsic not tagged",
+                "Add the #[pallet::call_index(...)] macro to the top of your extrinsic definition",
+                suggestion,
+                Applicability::MachineApplicable, // Suggestion can be applied automatically
+            );
+        }
+    }
+}
+
+//It checks if text above the function signature contains a call_index macro call.
```
Comment (KaiserKarel) [inline substrace_lints/src/substrace_lints/extrinsics_must_be_tagged.rs]:
```suggestion
// Checks if the text above the function signature contains a call_index macro call.
```

**Review by KaiserKarel — APPROVED** [review]

File: .github/workflows/build_test_clippy.yml (hunk 1 above)
Comment (TheCodingWombat) [reply .github/workflows/build_test_clippy.yml]: I merged main into this. Yes, still the same.

File: substrace_lints/src/substrace_lints/extrinsics_must_be_tagged.rs (hunk 3 above)
Comment (TheCodingWombat) [reply substrace_lints/src/substrace_lints/extrinsics_must_be_tagged.rs]: Suggestion applied.

File: Cargo.toml (hunk 2 above)
Comment (TheCodingWombat) [reply Cargo.toml]: Removed and still compiles.

**Review by KaiserKarel — APPROVED** [review]

**Pull request merged**